### PR TITLE
led_strip: add SPI backend (IEC-24)

### PR DESCRIPTION
**Pull request by Kainarx**
The interface still needs to be modified, the current version is only for technical verification

# Checklist

- [ ] Component contains License
- [ ] Component contains README.md
- [ ] Component contains idf_component.yml file with `url` field defined
- [ ] Component was added to [upload job](https://github.com/espressif/idf-extra-components/blob/master/.github/workflows/upload_component.yml#L18)
- [ ] Component was added to [build job](https://github.com/espressif/idf-extra-components/blob/master/test_app/CMakeLists.txt#L8)
- [ ] _Optional:_ Component contains unit tests
- [ ] CI passing

# Change description
Add an SPI mode to control Led_Strip by `CONFIG_BLINK_LED_SPI`

### CODE REVIEW

File: led_strip/examples/led_strip_rmt_ws2812/main/idf_component.yml [hunk 1]
```diff
@@ -3,3 +3,6 @@ dependencies:
   espressif/led_strip:
     version: '^2'
     override_path: '../../../'
+
+  led_strip:
+    path: /home/chenjichang/task/idf-extra-components/led_strip
```
Comment (suda-morris) [inline led_strip/examples/led_strip_rmt_ws2812/main/idf_component.yml:8]: Let's simplify the examples by creating a new example: `led_strip/examples/led_strip_spi_ws2812`. This line can be removed

File: led_strip/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,7 @@
+## 2.4.0
+- Support SPI mode `CONFIG_BLINK_LED_SPI`
```
Comment (suda-morris) [inline led_strip/CHANGELOG.md:2]: this config is not useable to the users. In fact, the SPI mode is always supported, they don't need to enable any Kconfig flags.

File: led_strip/include/led_strip.h [hunk 3]
```diff
@@ -8,7 +8,7 @@
 #include <stdint.h>
 #include "esp_err.h"
 #include "led_strip_rmt.h"
-
+#include "led_strip_spi.h"
```
Comment (suda-morris) [inline led_strip/include/led_strip.h:11]: nitpick, can we keep a blank line after this include statement?

File: led_strip/include/led_strip_spi.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+#include "soc/spi_periph.h"
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief LED Strip SPI specific configuration
+ */
+typedef struct {
+    rmt_clock_source_t clk_src; /*!< SPI clock source */
```
Comment (suda-morris) [inline led_strip/include/led_strip_spi.h:26]: spi_clock_source_t

Because this type was introduced since idf 5.1, so we need to change the dependency of this led_strip component as well

File: led_strip/include/led_strip_spi.h [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    uint32_t resolution_hz;     /*!< SPI tick resolution, if set to zero, a default resolution (10MHz) will be applied */
+    size_t mem_block_symbols;   /*!< How many RMT symbols can one RMT channel hold at one time. Set to 0 will fallback to use the default size. */
```
Comment (suda-morris) [inline led_strip/include/led_strip_spi.h:28]: why RMT here?

File: led_strip/include/led_strip_spi.h [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    spi_host_device_t host_channel ;
```
Comment (suda-morris) [inline led_strip/include/led_strip_spi.h:29]: unnecessary space before the ";". also suggest not using the word "channel". SPI doesn't have the concept of "channel". Here you may want to use "host controller"

File: led_strip/include/led_strip_spi.h [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    struct {
+        uint32_t with_dma: 1;   /*!< Use DMA to transmit data */
+    } flags;
+} led_strip_spi_config_t;
+
+/**
+ * @brief Create LED strip based on SPI MOSI channel
+ *
```
Comment (suda-morris) [inline led_strip/include/led_strip_spi.h:37]: please add some notes about this unusual SPI use. e.g. 

```
@note Although only the MOSI line is used for generating the signal, the whole SPI bus can't be used for other purposes.
```

File: led_strip/src/led_strip_spi_dev.c [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+#include <sys/cdefs.h>
+#include "esp_log.h"
+#include "esp_check.h"
+#include "driver/spi_master.h"
+#include "driver/spi_common.h"
+#include "led_strip.h"
+#include "led_strip_interface.h"
+#include "hal/spi_hal.h"
+
+#define LED_STRIP_SPI_DEFAULT_RESOLUTION (2.4 * 1000 * 1000) // 10MHz resolution
```
Comment (suda-morris) [inline led_strip/src/led_strip_spi_dev.c:17]: wrong comment?

Comment (suda-morris) [reply led_strip/src/led_strip_spi_dev.c:17]: and please note, because the SPI hardware doesn't have the fractional divisor, so this might not be the real resolution. Please call spi_device_get_actual_freq to get the final clock frequency.

File: led_strip/src/led_strip_spi_dev.c [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+//low_level:100 ,high_level:110
+#define LED_COLOR_BUF 3
+#define LED_BUF_BIT (LED_COLOR_BUF * 8)
+
+
+static const char *TAG = "led_strip_spi";
+
+typedef struct {
+    led_strip_t base;
+    spi_bus_config_t spi_bus;
```
Comment (suda-morris) [inline led_strip/src/led_strip_spi_dev.c:29]: if it's not used, we don't need to save it.

File: led_strip/src/led_strip_spi_dev.c [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+    if (led_config->led_pixel_format == LED_PIXEL_FORMAT_GRBW) {
+        bytes_per_pixel = 4;
+    } else if (led_config->led_pixel_format == LED_PIXEL_FORMAT_GRB) {
+        bytes_per_pixel = 3;
+    } else {
+        assert(false);
+    }
+    uint32_t mem_caps = MALLOC_CAP_DEFAULT;
+    if (spi_config->flags.with_dma) {
+        // DMA descriptors must be placed in internal SRAM
```
Comment (suda-morris) [inline led_strip/src/led_strip_spi_dev.c:138]:
```suggestion
        // DMA buffer should be placed in internal SRAM
```

File: led_strip/src/led_strip_spi_dev.c [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+        mem_caps |= MALLOC_CAP_32BIT | MALLOC_CAP_DMA;
+    }
+    spi_strip = heap_caps_calloc(1, sizeof(led_strip_spi_obj) + led_config->max_leds * bytes_per_pixel * LED_COLOR_BUF, mem_caps);
+
+    //spi_strip = calloc(1, sizeof(led_strip_spi_obj) + led_config->max_leds * bytes_per_pixel * LED_COLOR_BUF);
```
Comment (suda-morris) [inline led_strip/src/led_strip_spi_dev.c:143]: can we remove it?

File: led_strip/src/led_strip_spi_dev.c [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+    uint32_t resolution = spi_config->resolution_hz ? spi_config->resolution_hz : LED_STRIP_SPI_DEFAULT_RESOLUTION;
+
+    spi_strip->spi_host = spi_config->host_channel;
+    // for backward compatibility, if the user does not set the clk_src, use the default value
+    spi_clock_source_t clk_src = SPI_CLK_SRC_DEFAULT;
+    if (spi_config->clk_src) {
+        clk_src = spi_config->clk_src;
+    }
+
+    spi_bus_config_t spi_bus_cfg = {
```
Comment (suda-morris) [inline led_strip/src/led_strip_spi_dev.c:154]: please add some comments about why we set most of other io_num to `-1`

File: led_strip/src/led_strip_spi_dev.c [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+#if !CONFIG_IDF_TARGET_ESP32
+    //SPI_D_POL : This bit is used to set the idle polarity of MOSI. 1：high；0：low (ESP32 does not have this register, and its default polarity is low)
+    spi_dev_t *hw;
+    hw = SPI_LL_GET_HW(spi_strip->spi_host);
+    hw->ctrl.d_pol = 0;
+#endif
+
+
+    if (led_config->flags.invert_out == true) {
+        //Inverse the polarity of MOSI(not recommand)
```
Comment (suda-morris) [inline led_strip/src/led_strip_spi_dev.c:173]: why it's not recommended? if there's an inverter on the PCB (by a BJT or MOSFET), there user may need to enable this flag to invert it back in the software.

File: led_strip/src/led_strip_spi_dev.c [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+        esp_rom_gpio_connect_out_signal(led_config->strip_gpio_num, spi_periph_signal[spi_strip->spi_host].spid_out, true, false);
+    }
+
+    spi_device_interface_config_t spi_dev_cfg = {
+        .clock_source = clk_src,
+        .command_bits = 0,
+        .address_bits = 0,
+        .dummy_bits = 0,
+        .clock_speed_hz = resolution,
+        .duty_cycle_pos = 128,
```
Comment (suda-morris) [inline led_strip/src/led_strip_spi_dev.c:183]: please add comment on the meaning of setting it to `128`

File: led_strip/src/led_strip_spi_dev.c [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+        .mode = 0,
+        .spics_io_num = -1,
+        .queue_size = LED_STRIP_SPI_DEFAULT_TRANS_QUEUE_SIZE,
+    };
+
+    ESP_RETURN_ON_ERROR(spi_bus_add_device(spi_strip->spi_host, &spi_dev_cfg, &spi_strip->spi_device), TAG, "Failed to add spi device");
```
Comment (suda-morris) [inline led_strip/src/led_strip_spi_dev.c:189]: if return here, then `spi_strip` which was allocated before will be leaked forever.

File: led_strip/CHANGELOG.md (hunk 2 above)
Comment (suda-morris) [reply led_strip/CHANGELOG.md:2]: also please don't forget to bump the version in the idf_component.yml

File: led_strip/include/led_strip_spi.h [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+extern "C" {
+#endif
+
+/**
+ * @brief LED Strip SPI specific configuration
+ */
+typedef struct {
+    spi_clock_source_t clk_src; /*!< SPI clock source */
+    uint32_t resolution_hz;     /*!< SPI tick resolution, if set to zero, a default resolution (2.5MHz) will be applied */
+    spi_host_device_t spi_bus;
```
Comment (suda-morris) [inline led_strip/include/led_strip_spi.h:28]: missing API document

File: led_strip/src/led_strip_spi_dev.c (hunk 14 above)
Comment (wanckl) [reply led_strip/src/led_strip_spi_dev.c:182]: By default this one is 50% duty I think, and this one is for SPI CLK line, not effect for MOSI line, pls check really need this.

Comment (Kainarx) [reply led_strip/src/led_strip_spi_dev.c:182]: yes,this one is not related to MOSI, I will remove it

File: led_strip/CHANGELOG.md [hunk 17]
```diff
@@ -1,3 +1,9 @@
+## 2.4.0
+
+- Support configurable SPI mode to contorl leds
+  - The interface still needs to be modified, the current version is only for technical verification
```
Comment (suda-morris) [inline led_strip/CHANGELOG.md:4]: what do you mean by "needs to be modified"? which public function or structure do you see needs a modification in the sooner future?

File: led_strip/CHANGELOG.md [hunk 18]
```diff
@@ -1,3 +1,9 @@
...
+  - recommand to enable DMA when using SPI mode
```
Comment (suda-morris) [inline led_strip/CHANGELOG.md:5]: typo. `recommend`

File: led_strip/examples/led_strip_rmt_ws2812/main/led_strip_rmt_ws2812_main.c [hunk 19]
```diff
@@ -11,21 +11,23 @@
...
+#define LED_STRIP_BLINK_GPIO  14
 
 // LED numbers in the strip
-#define LED_STRIP_LED_NUMBERS 1
+#define LED_STRIP_LED_NUMBERS 24
 
-// 10MHz resolution, 1 tick = 0.1us (led strip needs a high resolution)
-#define LED_STRIP_RMT_RES_HZ  (10 * 1000 * 1000)
+//#define CONFIG_BLINK_LED_RMT
+#define CONFIG_BLINK_LED_SPI
```
Comment (suda-morris) [inline led_strip/examples/led_strip_rmt_ws2812/main/led_strip_rmt_ws2812_main.c:20]: shall we revert the changes in this file?

File: led_strip/include/led_strip_spi.h [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+#include <stdint.h>
+#include "esp_err.h"
+#include "driver/gpio.h"
+#include "esp_rom_gpio.h"
+#include "driver/spi_master.h"
+#include "driver/spi_common.h"
+#include "hal/spi_types.h"
+#include "led_strip_types.h"
+#include "soc/gpio_sig_map.h"
+#include "soc/spi_periph.h"
```
Comment (suda-morris) [inline led_strip/include/led_strip_spi.h:17]: Don't expose private header files in a public one.

```suggestion
#include "driver/spi_master.h"
#include "led_strip_types.h"
```

Comment (suda-morris) [reply led_strip/include/led_strip_spi.h:17]: you can include the private header files in the `.c` file

File: led_strip/examples/led_strip_spi_ws2812/main/idf_component.yml [hunk 21]
```diff
@@ -0,0 +1,6 @@
+## IDF Component Manager Manifest File
+dependencies:
+  espressif/led_strip:
+    version: '^2'
```
Comment (suda-morris) [inline led_strip/examples/led_strip_spi_ws2812/main/idf_component.yml:4]:
```suggestion
    version: '^2.4'
```

File: led_strip/examples/led_strip_spi_ws2812/main/led_strip_spi_ws2812_main.c [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+//#define CONFIG_BLINK_LED_RMT
+#define CONFIG_BLINK_LED_SPI
+
+static const char *TAG = "example";
+
+#define LED_STRIP_SPI_RES_HZ  (2.5 * 1000 * 1000)
+// 2.5MHz resolution, 1 tick = 0.4us (led strip needs a high resolution)
+
+// LED Strip object handle
+led_strip_handle_t led_strip;
```
Comment (suda-morris) [inline led_strip/examples/led_strip_spi_ws2812/main/led_strip_spi_ws2812_main.c:28]: not a fan of using global variables. why not return the led_strip handle by the `configure_led` function?

File: led_strip/examples/led_strip_spi_ws2812/main/led_strip_spi_ws2812_main.c [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+
+void configure_led()
```
Comment (suda-morris) [inline led_strip/examples/led_strip_spi_ws2812/main/led_strip_spi_ws2812_main.c:30]: if a function doesn't need any parameters, please use `void`

File: led_strip/examples/led_strip_spi_ws2812/main/led_strip_spi_ws2812_main.c [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+        .led_pixel_format = LED_PIXEL_FORMAT_GRB, // Pixel format of your LED strip
+        .led_model = LED_MODEL_WS2812,            // LED strip model
+        .flags.invert_out = false,                // whether to invert the output signal
+    };
+
+    // LED strip backend configuration: SPI
+    led_strip_spi_config_t spi_config = {
+        .clk_src = SPI_CLK_SRC_DEFAULT,        // different clock source can lead to different power consumption
+        .resolution_hz = LED_STRIP_SPI_RES_HZ, // SPI counter clock frequency
+        .flags.with_dma = true,               // DMA feature is available on ESP target like ESP32-S3
```
Comment (suda-morris) [inline led_strip/examples/led_strip_spi_ws2812/main/led_strip_spi_ws2812_main.c:45]: the comments seems like copied from the RMT example... In our chip, SPI peripheral always support DMA and non-DMA. But we recommend to use the DMA in the led_strip application. Please update the comments accordingly.

File: led_strip/src/led_strip_spi_dev.c [hunk 25]
```diff
@@ -0,0 +1,212 @@
...
+    //SPI_D_POL : This bit is used to set the idle polarity of MOSI. 1：high；0：low (ESP32 does not have this register, and its default polarity is low)
+    spi_dev_t *hw;
+    hw = SPI_LL_GET_HW(spi_strip->spi_host);
+    hw->ctrl.d_pol = 0;
+#endif
+
+
+    if (led_config->flags.invert_out == true) {
+        esp_rom_gpio_connect_out_signal(led_config->strip_gpio_num, spi_periph_signal[spi_strip->spi_host].spid_out, true, false);
+    }
```
Comment (wanckl) [inline led_strip/src/led_strip_spi_dev.c:173]: What the relationship between above two code to inverse the default MOSI level ?   
And :
> and its default polarity is low   

default level of ESP32 shouldn't be low, isn't it because your inverse it by 'and its default polarity is low' ?

Comment (wanckl) [reply led_strip/src/led_strip_spi_dev.c:173]: Sorry to find that default level of MOSI pin of esp32 is really low, and different from all other chips now.    

But as SPI protocol don't have a standard for data line (MOSI and MISO) by default, driver don't need to guarantee this, you can set default level in your application if needed.

File: led_strip/examples/led_strip_rmt_ws2812/main/led_strip_rmt_ws2812_main.c [hunk 26]
```diff
@@ -11,21 +11,17 @@
...
 // LED numbers in the strip
-#define LED_STRIP_LED_NUMBERS 1
-
-// 10MHz resolution, 1 tick = 0.1us (led strip needs a high resolution)
-#define LED_STRIP_RMT_RES_HZ  (10 * 1000 * 1000)
+#define LED_STRIP_LED_NUMBERS 24
 
 static const char *TAG = "example";
+#define LED_STRIP_RMT_RES_HZ  (10 * 1000 * 1000)
+// 10MHz resolution, 1 tick = 0.1us (led strip needs a high resolution)
```
Comment (suda-morris) [inline led_strip/examples/led_strip_rmt_ws2812/main/led_strip_rmt_ws2812_main.c:21]: reorder the definition and comments
```suggestion
// Numbers of the LED in the strip
#define LED_STRIP_LED_NUMBERS 24
// 10MHz resolution, 1 tick = 0.1us (led strip needs a high resolution)
#define LED_STRIP_RMT_RES_HZ  (10 * 1000 * 1000)

static const char *TAG = "example";
```

File: led_strip/include/led_strip_spi.h [hunk 27]
```diff
@@ -0,0 +1,47 @@
+/*
+ * SPDX-FileCopyrightText: 2022-2023 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#pragma once
+
+#include <stdint.h>
+#include "esp_err.h"
+#include "hal/spi_types.h"
```
Comment (suda-morris) [inline led_strip/include/led_strip_spi.h:10]: Better to rely on the driver, not the hal.

```suggestion
#include "driver/spi_master.h"
```

File: led_strip/examples/led_strip_spi_ws2812/main/led_strip_spi_ws2812_main.c [hunk 28]
```diff
@@ -0,0 +1,75 @@
...
+// GPIO assignment
+#define LED_STRIP_BLINK_GPIO  14
+
+// LED numbers in the strip
+#define LED_STRIP_LED_NUMBERS 24
+
+static const char *TAG = "example";
+
+#define LED_STRIP_SPI_RES_HZ  (2.5 * 1000 * 1000)
+// 2.5MHz resolution, 1 tick = 0.4us (led strip needs a high resolution)
```
Comment (suda-morris) [inline led_strip/examples/led_strip_spi_ws2812/main/led_strip_spi_ws2812_main.c:22]:
```suggestion
#define LED_STRIP_LED_NUMBERS 24
// 2.5MHz resolution, 1 tick = 0.4us (led strip needs a high resolution)
#define LED_STRIP_SPI_RES_HZ  (2.5 * 1000 * 1000)

static const char *TAG = "example";
```

File: led_strip/src/led_strip_spi_dev.c [hunk 29]
```diff
@@ -0,0 +1,213 @@
...
+#include <sys/cdefs.h>
+#include "esp_log.h"
+#include "esp_check.h"
+#include "esp_rom_gpio.h"
+#include "driver/spi_master.h"
+#include "driver/spi_common.h"
+#include "soc/spi_periph.h"
+#include "led_strip.h"
+#include "led_strip_interface.h"
+#include "hal/spi_hal.h"
```
Comment (suda-morris) [inline led_strip/src/led_strip_spi_dev.c:17]: nitpick, this one is not necessary, it's already covered by the "driver/spi_master.h"

File: led_strip/src/led_strip_spi_dev.c [hunk 30]
```diff
@@ -0,0 +1,213 @@
...
+
+
+#define LED_STRIP_SPI_DEFAULT_RESOLUTION (2.5 * 1000 * 1000) // 2.5MHz resolution
+#define LED_STRIP_SPI_DEFAULT_TRANS_QUEUE_SIZE 4
+//Each color of 1 bit is represented by 3 bits, a color occupies 24 bits, 3 byte
```
Comment (suda-morris) [inline led_strip/src/led_strip_spi_dev.c:22]: 3bits of what?

File: led_strip/src/led_strip_spi_dev.c [hunk 31]
```diff
@@ -0,0 +1,213 @@
...
+//low_level:100 ,high_level:110
+#define LED_COLOR_BUF 3
```
Comment (suda-morris) [inline led_strip/src/led_strip_spi_dev.c:24]: the name is confusing. maybe you can use `SPI_BITS_PER_COLOR_BIT`

File: led_strip/src/led_strip_spi_dev.c (hunk 30 above)
Comment (suda-morris) [reply led_strip/src/led_strip_spi_dev.c:22]: and a color is not always 24bits. If the `led_pixel_format_t` is set to `LED_PIXEL_FORMAT_GRBW`, then it's 32bits for each color

File: led_strip/src/led_strip_spi_dev.c [hunk 32]
```diff
@@ -0,0 +1,213 @@
...
+#define LED_BUF_BIT (LED_COLOR_BUF * 8)
```
Comment (suda-morris) [inline led_strip/src/led_strip_spi_dev.c:25]:
```suggestion
#define SPI_BITS_PER_COLOR_ELEMENT (SPI_BITS_PER_COLOR_BIT * 8)
```

File: led_strip/src/led_strip_spi_dev.c [hunk 33]
```diff
@@ -0,0 +1,213 @@
...
+#include "soc/spi_periph.h"
+#include "led_strip.h"
+#include "led_strip_interface.h"
+#include "hal/spi_hal.h"
+
+
+#define LED_STRIP_SPI_DEFAULT_RESOLUTION (2.5 * 1000 * 1000) // 2.5MHz resolution
+#define LED_STRIP_SPI_DEFAULT_TRANS_QUEUE_SIZE 4
+//Each color of 1 bit is represented by 3 bits, a color occupies 24 bits, 3 byte
+//low_level:100 ,high_level:110
```
Comment (suda-morris) [inline led_strip/src/led_strip_spi_dev.c:23]: shall we put this comment in the `__led_strip_spi_bit` function?

File: led_strip/src/led_strip_spi_dev.c [hunk 34]
```diff
@@ -0,0 +1,212 @@
...
+    *(buf + 0) |= data & BIT(6) ? BIT(4) | BIT(3) : BIT(4);
+    *(buf + 0) |= data & BIT(7) ? BIT(7) | BIT(6) : BIT(7);
+}
+
+static esp_err_t led_strip_spi_set_pixel(led_strip_t *strip, uint32_t index, uint32_t red, uint32_t green, uint32_t blue)
+{
+    led_strip_spi_obj *spi_strip = __containerof(strip, led_strip_spi_obj, base);
+    ESP_RETURN_ON_FALSE(index < spi_strip->strip_len, ESP_ERR_INVALID_ARG, TAG, "index out of maximum number of LEDs");
+    uint32_t start = index * spi_strip->bytes_per_pixel * LED_COLOR_BUF;
+    memset(spi_strip->pixel_buf + start, 0, spi_strip->bytes_per_pixel * LED_COLOR_BUF);
```
[the remark this answers is not in the document]
Comment (suda-morris) [reply led_strip/src/led_strip_spi_dev.c:58]: what if the `spi_strip->bytes_per_pixel > 3`

File: led_strip/src/led_strip_spi_dev.c [hunk 35]
```diff
@@ -0,0 +1,213 @@
...
+    led_strip_spi_obj *spi_strip = __containerof(strip, led_strip_spi_obj, base);
+    ESP_RETURN_ON_FALSE(index < spi_strip->strip_len, ESP_ERR_INVALID_ARG, TAG, "index out of maximum number of LEDs");
+    ESP_RETURN_ON_FALSE(spi_strip->bytes_per_pixel == 4, ESP_ERR_INVALID_ARG, TAG, "wrong LED pixel format, expected 4 bytes per pixel");
+    uint32_t start = index * spi_strip->bytes_per_pixel * LED_COLOR_BUF;
+    // SK6812 component order is GRBW
+    memset(spi_strip->pixel_buf + start, 0, spi_strip->bytes_per_pixel * LED_COLOR_BUF);
+    __led_strip_spi_bit(green, &spi_strip->pixel_buf[start]);
+    __led_strip_spi_bit(red, &spi_strip->pixel_buf[start + LED_COLOR_BUF]);
+    __led_strip_spi_bit(blue, &spi_strip->pixel_buf[start + LED_COLOR_BUF * 2]);
+    __led_strip_spi_bit(blue, &spi_strip->pixel_buf[start + LED_COLOR_BUF * 3]);
```
Comment (suda-morris) [inline led_strip/src/led_strip_spi_dev.c:77]: why blue again???

File: led_strip/src/led_strip_spi_dev.c [hunk 36]
```diff
@@ -0,0 +1,213 @@
...
+    // Write zero to turn off all leds
+    if (spi_strip->bytes_per_pixel == 3) {
+        for (int index = 0; index < spi_strip->strip_len; index++) {
+            led_strip_spi_set_pixel(strip, index, 0, 0, 0);
+        }
+    } else if (spi_strip->bytes_per_pixel == 4) {
+        for (int index = 0; index < spi_strip->strip_len; index++) {
+            led_strip_spi_set_pixel_rgbw(strip, index, 0, 0, 0, 0);
+        }
+    }
```
Comment (suda-morris) [inline led_strip/src/led_strip_spi_dev.c:108]: I think this can be cimplified by calling  `__led_strip_spi_bit`  for each byte in the `pixel_buf`?

File: led_strip/src/led_strip_spi_dev.c [hunk 37]
```diff
@@ -0,0 +1,213 @@
...
+        bytes_per_pixel = 4;
+    } else if (led_config->led_pixel_format == LED_PIXEL_FORMAT_GRB) {
+        bytes_per_pixel = 3;
+    } else {
+        assert(false);
+    }
+    uint32_t mem_caps = MALLOC_CAP_DEFAULT;
+    if (spi_config->flags.with_dma) {
+        // DMA buffer must be placed in internal SRAM
+        mem_caps |= MALLOC_CAP_32BIT | MALLOC_CAP_DMA;
```
Comment (suda-morris) [inline led_strip/src/led_strip_spi_dev.c:141]: why it has to be aligned to 32bits? instead of 8bits?

File: led_strip/src/led_strip_spi_dev.c [hunk 38]
```diff
@@ -0,0 +1,213 @@
...
+    }
+    spi_strip = heap_caps_calloc(1, sizeof(led_strip_spi_obj) + led_config->max_leds * bytes_per_pixel * LED_COLOR_BUF, mem_caps);
```
Comment (suda-morris) [inline led_strip/src/led_strip_spi_dev.c:143]: Okay, here the `LED_COLOR_BUF` has a different meaning, it's the factor between the "one color byte" and the "spi transfer bytes". heh naming is always the hardest thing in programming.

File: led_strip/src/led_strip_spi_dev.c (hunk 37 above)
Comment (wanckl) [reply led_strip/src/led_strip_spi_dev.c:141]: This one is restricted by DMA I think

Comment (Kainarx) [reply led_strip/src/led_strip_spi_dev.c:141]: yes, here 8bits is ok, buffer which is in internal RAM need to be aligned to 32bits only in burst and inlink mode

File: led_strip/CMakeLists.txt [hunk 39]
```diff
@@ -4,6 +4,16 @@ if(CONFIG_SOC_RMT_SUPPORTED)
...
+
+list(APPEND srcs "src/led_strip_spi_dev.c")
+
+# Add here if the component is compatible with IDF >= v5.1
+if("${IDF_VERSION_MAJOR}.${IDF_VERSION_MINOR}" VERSION_GREATER_EQUAL "5.1")
+
+    list(APPEND EXTRA_COMPONENT_DIRS ../bdc_motor ../led_strip ../sh2lib ../nghttp ../esp_serial_slave_link ../onewire_bus ../ccomp_timer)
+
+endif()
+
```
Comment (suda-morris) [inline led_strip/CMakeLists.txt:16]: these changes should be made to https://github.com/espressif/idf-extra-components/blob/master/test_app/CMakeLists.txt

File: test_app/CMakeLists.txt [hunk 40]
```diff
@@ -14,7 +14,12 @@ endif()
 
 # 3. Add here if the component is compatible with IDF >= v5.0
 if("${IDF_VERSION_MAJOR}.${IDF_VERSION_MINOR}" VERSION_GREATER_EQUAL "5.0")
-    list(APPEND EXTRA_COMPONENT_DIRS ../bdc_motor ../led_strip ../sh2lib ../nghttp ../esp_serial_slave_link ../onewire_bus ../ccomp_timer)
+    list(APPEND EXTRA_COMPONENT_DIRS ../bdc_motor ../sh2lib ../nghttp ../esp_serial_slave_link ../onewire_bus ../ccomp_timer)
+endif()
+
+# 3. Add here if the component is compatible with IDF >= v5.1
```
Comment (suda-morris) [inline test_app/CMakeLists.txt:20]:
```suggestion
# 4. Add here if the component is compatible with IDF >= v5.1
```

File: led_strip/src/led_strip_rmt_dev.c [hunk 41]
```diff
@@ -37,6 +37,7 @@ static esp_err_t led_strip_rmt_set_pixel(led_strip_t *strip, uint32_t index, uin
 {
     led_strip_rmt_obj *rmt_strip = __containerof(strip, led_strip_rmt_obj, base);
     ESP_RETURN_ON_FALSE(index < rmt_strip->strip_len, ESP_ERR_INVALID_ARG, TAG, "index out of maximum number of LEDs");
+    ESP_RETURN_ON_FALSE(rmt_strip->bytes_per_pixel == 3, ESP_ERR_INVALID_ARG, TAG, "wrong LED pixel format, expected 3 bytes per pixel");
```
Comment (suda-morris) [inline led_strip/src/led_strip_rmt_dev.c:40]: the previous logic doesn't forbid the user from using this function on an RGBW format led strip (check the code below, if the white element is not specified. we set it to zero).

Comment (suda-morris) [reply led_strip/src/led_strip_rmt_dev.c:40]: This logic should also be applied to the spi interfaced led strip.

File: led_strip/examples/led_strip_spi_ws2812/main/led_strip_spi_ws2812_main.c [hunk 42]
```diff
@@ -0,0 +1,73 @@
...
+
+    led_strip_handle_t led_strip = configure_led();
+    bool led_on_off = false;
+
+    ESP_LOGI(TAG, "Start blinking LED strip");
+    while (1) {
+        if (led_on_off) {
+            /* Set the LED pixel using RGB from 0 (0%) to 255 (100%) for each color */
+            for (int i = 0; i < LED_STRIP_LED_NUMBERS; i++) {
+                ESP_ERROR_CHECK(led_strip_set_pixel(led_strip, i, i, 7, 5));
```
Comment (suda-morris) [inline led_strip/examples/led_strip_spi_ws2812/main/led_strip_spi_ws2812_main.c:59]: why the **red** element should be assigned with `i` while other elements use a const value?

Comment (Kainarx) [reply led_strip/examples/led_strip_spi_ws2812/main/led_strip_spi_ws2812_main.c:59]: it's just for test , I will set it to const value later

**Comment by suda-morris** [comment]
https://github.com/espressif/idf-extra-components/pull/194 was just merged. @Kainarx Please rebase your branch to the lastest **upstream**'s master branch in order to make the CI pass.

File: led_strip/src/led_strip_spi_dev.c [hunk 43]
```diff
@@ -0,0 +1,219 @@
...
+
+
+static void __led_strip_spi_bit(uint8_t data, uint8_t *buf)
+{
+    // Each color of 1 bit is represented by 3 bits of SPI, low_level:100 ,high_level:110
+    // So a color byte occupies 3 bytes of SPI.
+    *(buf + 2) |= data & BIT(0) ? BIT(2) | BIT(1) : BIT(2);
+    *(buf + 2) |= data & BIT(1) ? BIT(5) | BIT(4) : BIT(5);
+    *(buf + 2) |= data & BIT(2) ? BIT(7) : 0x00;
+    *(buf + 1) |= data & BIT(2) ? BIT(0) : BIT(0);
```
Comment (suda-morris) [inline led_strip/src/led_strip_spi_dev.c:42]: simplified to 
```suggestion
    *(buf + 1) |= BIT(0);
```

File: led_strip/src/led_strip_spi_dev.c [hunk 44]
```diff
@@ -0,0 +1,219 @@
...
+    }
+    uint32_t mem_caps = MALLOC_CAP_DEFAULT;
+    if (spi_config->flags.with_dma) {
+        // DMA buffer must be placed in internal SRAM
+        mem_caps |= MALLOC_CAP_INTERNAL | MALLOC_CAP_DMA;
+    }
+    spi_strip = heap_caps_calloc(1, sizeof(led_strip_spi_obj) + led_config->max_leds * bytes_per_pixel * SPI_BYTES_PER_COLOR_BYTE, mem_caps);
+
+    ESP_GOTO_ON_FALSE(spi_strip, ESP_ERR_NO_MEM, err, TAG, "no mem for spi strip");
+    uint32_t resolution = spi_config->resolution_hz ? spi_config->resolution_hz : LED_STRIP_SPI_DEFAULT_RESOLUTION;
```
Comment (suda-morris) [inline led_strip/src/led_strip_spi_dev.c:144]: this can be defined just before where it's used.

**Comment by suda-morris** [comment]
@Kainarx Nice work! I'm going to merge this PR now. I will create a new PR later to fix these nitpicks.

**Review by suda-morris — APPROVED** [review]
